src/core/thread/dua_manager.hpp

```diff
@@ -129,17 +149,118 @@ class DuaManager : public InstanceLocator
      *
      */
     void Restore(void);
+#endif
+
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+    void UpdateChildDomainUnicastAddress(const Child &aChild, Mle::ChildDuaState aState);
+#endif
 
 private:
+    enum
+    {
+        kNewRouterRegistrationDelay = 5,    ///< Delay (in seconds) for waiting link establishment for a new Router.
+        kStateUpdatePeriod          = 1000, ///< 1000ms period  (i.e. 1s)
+    };
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
     otError GenerateDomainUnicastAddressIid(void);
     otError Store(void);
 
+    void AddDomainUnicastAddress(void);
+    void RemoveDomainUnicastAddress(void);
+    void UpdateRegistrationDelay(uint8_t aDelay);
+#endif
+
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+    void SendAddressNotification(Ip6::Address &aAddress, ThreadStatusTlv::DuaStatus aStatus, const Child &aChild);
+#endif
+
+    static void HandleNotifierEvents(Notifier::Receiver &aReceiver, Events aEvents)
+    {
+        static_cast<DuaManager &>(aReceiver).HandleNotifierEvents(aEvents);
+    }
+
+    void HandleNotifierEvents(Events aEvents);
+
+    static void HandleTimer(Timer &aTimer) { aTimer.GetOwner<DuaManager>().HandleTimer(); }
+
+    void HandleTimer(void);
+
+    static void HandleRegistrationTask(Tasklet &aTasklet) { aTasklet.GetOwner<DuaManager>().PerformNextRegistration(); }
```

Comment:
Not specifically related to this PR, but just want to have more a general discussion on the static callback methods.

Recently we are moving towards defining them in header file instead of the `cpp` file.

I see little benefit in doing this and actually may be some limitations. I suggest we define them in `cpp` file.

- These `static` methods are used as callbacks and we use their address (as a function pointer) passed as callback to some other object. So they cannot be inlined, so having them in the header should not help with toolchain inlining them.
- With the ones that use `GetOwner<>()` method, we may hinder the toolchain inline optimization (with some older toolchains). An older toolchain may not be able to inline the template method implementation if it is not visible/known at the point that it's being used.
   - Note that `locator.hpp` defines the template method `GetOwner<>` without an implementation
   - And the implementation is in `locator-getters.hpp`.
   - The idea is that OT core header files `#include <common/locator.hpp>` and then the `cpp` files `#include <common/locator-getters.hpp>`
   - This is to avoid the loop dependencies of header files. Note that `locator-getters.hpp` needs to include `instance.hpp` which itself includes every other OT core header file. So we cannot directly use `locator-getters.hpp` from other OT core header files.

Thoughts?

Reply:
Thanks for the details. SGTM. We may have a separate PR specially for this.